java-checks/src/main/java/org/sonar/java/checks/UndocumentedApiCheck.java

```diff
@@ -298,8 +317,87 @@ private boolean hasNonVoidReturnType(Tree tree) {
     return false;
   }
 
-  private static boolean hasReturnJavadoc(String comment) {
-    return comment.contains("@return");
+  private static class Javadoc {
+    private static final Pattern PARAMETER_JAVADOC_PATTERN = Pattern.compile(".*@param\\s++(?<name>\\S*)\\s++(?<descr>.+)");
+    private static final Pattern EXCEPTION_JAVADOC_PATTERN = Pattern.compile(".*@throws\\s++(?<name>\\S*)\\s++(?<descr>.+)");
+    private static final Pattern RETURN_JAVADOC_PATTERN = Pattern.compile(".*@return\\s++(?<descr>.+)");
+    private static final Set<String> PLACEHOLDERS = ImmutableSet.of("TODO", "FIXME", "...", ".");
+
+    final String mainDescription;
+    final Map<String, String> parameters;
+    final Map<String, String> thrownExceptions;
+    final String returnDescription;
+
+    Javadoc(@Nullable String javadoc) {
+      List<String> lines = lines(javadoc);
+      mainDescription = extractMainDescription(lines);
+      parameters = extractToMap(lines, PARAMETER_JAVADOC_PATTERN);
+      thrownExceptions = extractToMap(lines, EXCEPTION_JAVADOC_PATTERN);
+      returnDescription = extractReturnDescription(lines);
+    }
+
+    private List<String> getUndocumentedParameters(List<String> parameterNames) {
+      return parameterNames.stream().filter(name -> Javadoc.isEmptyDescription(parameters.get(name))).collect(Collectors.toList());
+    }
+
+    private List<String> getUndocumentedThrownExceptions(List<String> exceptionNames) {
+      return exceptionNames.stream().filter(name -> Javadoc.isEmptyDescription(thrownExceptions.get(name))).collect(Collectors.toList());
+    }
+
+    private static boolean isEmptyDescription(@Nullable String part) {
+      return part == null
+        || part.trim().isEmpty()
+        || PLACEHOLDERS.contains(part)
+        || (!part.contains("{@inheritDoc}") && notEnoughWords(part));
+    }
+
+    private static boolean notEnoughWords(String javadoc) {
+      return javadoc.split("\\s").length < 2;
+    }
+
+    private static List<String> lines(@Nullable String javadoc) {
+      if (javadoc == null) {
+        return Collections.emptyList();
+      }
+      String cleanupJavaDoc = cleanupJavaDoc(javadoc);
+      String[] lines = cleanupJavaDoc.split("\\r?\\n");
+      return Arrays.stream(lines).map(String::trim).collect(Collectors.toList());
+    }
+
+    private static String cleanupJavaDoc(String javadoc) {
+      return javadoc.trim().substring(3).replace("*/", "").replace("*", "").trim();
```

Comment:
The replace "*" by "" seems a bit strong : you are going to strip any * character from the javadoc. Given what is verified I think it is fine but that wouldn't it deserve a comment ?